core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/StructuredRecordFileInputFormat.java

```diff
@@ -0,0 +1,146 @@
+      }
+      return (FileInputFormat<?, ?>) inputFormatObj;
+    } catch (ClassNotFoundException e) {
+      throw new IOException(String.format("Input format class %s could not be found.", className), e);
+    } catch (InstantiationException e) {
+      throw new IOException(String.format("Input format class %s could not be instantiated.", className), e);
+    } catch (IllegalAccessException e) {
+      throw new IOException(String.format("Illegal access while instantiating input format class %s.", className), e);
+    }
+  }
+
+  /**
+   * Record reader that tries to get the path each key-value was read from by looking at the input split.
+   */
+  public static class StructuredRecordReader extends RecordReader<NullWritable, StructuredRecord> {
+    private String path;
+    private RecordReader delegate;
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+      if (split instanceof FileSplit) {
+        path = ((FileSplit) split).getPath().toUri().toString();
+      } else if (split instanceof CombineFileSplit) {
+        path = Joiner.on(',').join(((CombineFileSplit) split).getPaths());
+      } else {
+        path = null;
+      }
+      Configuration conf = context.getConfiguration();
+      String className = conf.get(INPUT_FORMAT_CLASS);
+      try {
+        Class<?> inputFormatClass = conf.getClassByName(className);
+        Object inputFormatObj = inputFormatClass.newInstance();
+        if (!(inputFormatObj instanceof InputFormat)) {
+          throw new IOException(String.format("%s is not an input format", className));
+        }
+        InputFormat inputFormat = (InputFormat) inputFormatObj;
+        delegate = inputFormat.createRecordReader(split, context);
+        delegate.initialize(split, context);
+      } catch (ClassNotFoundException e) {
+        throw new IOException(String.format("Input format class %s could not be found.", className), e);
+      } catch (InstantiationException e) {
+        throw new IOException(String.format("Input format class %s could not be instantiated.", className), e);
+      } catch (IllegalAccessException e) {
+        throw new IOException(String.format("Illegal access while instantiating input format class %s.", className), e);
+      }
+    }
+
+    @Override
+    public boolean nextKeyValue() throws IOException, InterruptedException {
+      return delegate.nextKeyValue();
+    }
+
+    @Override
+    public NullWritable getCurrentKey() throws IOException, InterruptedException {
+      return NullWritable.get();
+    }
+
+    @Override
+    public StructuredRecord getCurrentValue() throws IOException, InterruptedException {
+      Object key = delegate.getCurrentKey();
+      Object val = delegate.getCurrentValue();
+      return StructuredRecord.builder(SCHEMA)
+        .set("path", path)
+        .set("offset", key instanceof LongWritable ? ((LongWritable) key).get() : null)
```

Comment:
Is the offset, the offset within that particular file or is it offset across all files put together?